## Week 3  Data Wrangling with NumPy and Pandas
This notebook demonstrates NumPy fundamentals and Pandas data wrangling techniques.
here we have to create and work with NumPy arrays, Load and explore a public CSV dataset, Handle the  missing values, Remove duplicate records, Filter data, Group data using groupby(), Merging datasets, Transforming data, Creating an analysis-ready dataset

### Importing NumPy and Pandas

In [9]:
import numpy as np
import pandas as pd
import os


## 1. NumPy Fundamentals

NumPy provides arrays and numerical operations that are useful for working with numerical data.
In this section, array creation, indexing, broadcasting, and vectorized operations will be practiced.

In [4]:
numbers = np.array([10, 20, 30, 40, 50])

print("NumPy array:", numbers)
print("Array type:", type(numbers))

NumPy array: [10 20 30 40 50]
Array type: <class 'numpy.ndarray'>


### NumPy Array Indexing

Indexing is used to access individual elements from a NumPy array.
Python uses zero-based indexing, so the first element has index 0.

In [5]:
print("First element:", numbers[0])
print("Third element:", numbers[2])
print("Last element:", numbers[-1])

First element: 10
Third element: 30
Last element: 50


### NumPy Broadcasting

Broadcasting allows a value to be applied to every element of a NumPy array without writing a separate loop.

In [6]:
increased_numbers = numbers + 5

print("Original array:", numbers)
print("After adding 5:", increased_numbers)

Original array: [10 20 30 40 50]
After adding 5: [15 25 35 45 55]


### NumPy Vectorized Operations

Vectorized operations allow calculations to be performed on all elements of an array at the same time.

In [7]:
prices = np.array([100, 200, 300, 400, 500])

discounted_prices = prices * 0.90

print("Original prices:", prices)
print("Prices after 10% discount:", discounted_prices)

Original prices: [100 200 300 400 500]
Prices after 10% discount: [ 90. 180. 270. 360. 450.]


## locating the CSV file path 

In [13]:
print(os.listdir())

['.ipynb_checkpoints', 'data', 'Week_03_Data_Wrangling.ipynb']


In [14]:
print(os.listdir("data"))

['Titanic-Dataset.csv']


In [15]:
df = pd.read_csv("data/Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [16]:
df = pd.read_csv("data/Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Checking the detaset size

In [17]:
print("Number of rows and columns:", df.shape)

Number of rows and columns: (891, 12)


## Exploring the Dataset Structure

The structure of the dataset is examined to identify the available columns, data types, and number of non-null values.

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Checking the column Name 

In [19]:
print("Dataset columns:")
print(df.columns.tolist())

Dataset columns:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


### Generating the basic Statics

In [20]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## Checking Missing Values

Missing values are checked before cleaning the dataset.

Identifying missing values helps determine which columns require data cleaning or treatment.

In [21]:
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


### Missing Value Percentage

The percentage of missing values in each column is calculated to understand the extent of missing data.

In [22]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

print("Missing value percentage:")
print(missing_percentage)

Missing value percentage:
PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64


## Data Cleaning

The raw dataset is kept unchanged, and a copy is created for the cleaning process.The missing values in the dataset will be handled using appropriate methods based on the type and amount of missing data.

In [23]:
cleaned_df = df.copy()

print("Original dataset shape:", df.shape)
print("Cleaning dataset shape:", cleaned_df.shape)

Original dataset shape: (891, 12)
Cleaning dataset shape: (891, 12)


### Handling Missing Age Values

The Age column contains missing numerical values.The median age is used to fill the missing values because the median is less affected by extreme values than the mean.

In [24]:
age_median = cleaned_df["Age"].median()

cleaned_df["Age"] = cleaned_df["Age"].fillna(age_median)

print("Median age used:", age_median)
print("Missing Age values after cleaning:", cleaned_df["Age"].isnull().sum())

Median age used: 28.0
Missing Age values after cleaning: 0


### Handling Missing Embarked Values

The Embarked column is categorical and contains a small number of missing values.The missing values are replaced with the most frequently occurring value, known as the mode.

In [25]:
embarked_mode = cleaned_df["Embarked"].mode()[0]

cleaned_df["Embarked"] = cleaned_df["Embarked"].fillna(embarked_mode)

print("Mode used:", embarked_mode)
print("Missing Embarked values after cleaning:", cleaned_df["Embarked"].isnull().sum())

Mode used: S
Missing Embarked values after cleaning: 0


### Handling the Cabin Column

The Cabin column contains approximately 77% missing values. Because most records do not contain cabin information, filling such a large proportion of the column could introduce unreliable information. Therefore, the Cabin column is removed from the cleaned analysis-ready dataset.

In [26]:
cleaned_df = cleaned_df.drop(columns=["Cabin"])

print("Cabin column removed.")

Cabin column removed.


### Checking Missing Values After Cleaning

The dataset is checked again to confirm that the missing values have been handled.

In [27]:
print("Missing values after cleaning:")
print(cleaned_df.isnull().sum())

Missing values after cleaning:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


## Checking and Removing Duplicate Records

Duplicate records can cause the same information to be counted more than once. The dataset is checked for duplicate rows before removing them.

In [28]:
duplicate_count = cleaned_df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


## Filtering Data

Filtering is used to select specific rows from a dataset based on a condition. In this section, passengers from a particular passenger class will be selected from the cleaned dataset.

In [29]:
first_class = cleaned_df[cleaned_df["Pclass"] == 1]

print("Number of first-class passengers:", len(first_class))

first_class.head()

Number of first-class passengers: 216


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,S
23,24,1,1,"Sloper, Mr. William Thompson",male,28.0,0,0,113788,35.5000,S


### Filtering Passengers by Survival Status

The dataset can also be filtered to select passengers who survived.

In [30]:
survived_passengers = cleaned_df[cleaned_df["Survived"] == 1]

print("Number of passengers who survived:", len(survived_passengers))

survived_passengers.head()

Number of passengers who survived: 342


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C


### Filtering Using Multiple Conditions

Multiple conditions can be combined to obtain more specific records. The following example selects female passengers who were in first class.

In [31]:
female_first_class = cleaned_df[
    (cleaned_df["Sex"] == "female") &
    (cleaned_df["Pclass"] == 1)
]

print("Number of female first-class passengers:", len(female_first_class))

female_first_class.head()

Number of female first-class passengers: 94


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,28.0,1,0,PC 17569,146.5208,C
52,53,1,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",female,49.0,1,0,PC 17572,76.7292,C


##  Grouping Data Using groupby()

The groupby() function is used to divide data into groups based on a particular column and calculate summary information for each group. In this section, passenger survival will be summarized by passenger class.

In [32]:
survival_by_class = cleaned_df.groupby("Pclass")["Survived"].mean()

print("Average survival rate by passenger class:")
print(survival_by_class)

Average survival rate by passenger class:
Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


### Grouping Survival by Sex

The survival rate can also be grouped by sex to compare the values between male and female passengers.

In [33]:
survival_by_sex = cleaned_df.groupby("Sex")["Survived"].mean()

print("Average survival rate by sex:")
print(survival_by_sex)

Average survival rate by sex:
Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


### Grouping by Multiple Columns
Pandas can group data using more than one column. The following example calculates the average survival rate for each combination of passenger class and sex.

In [34]:
survival_by_class_and_sex = cleaned_df.groupby(
    ["Pclass", "Sex"]
)["Survived"].mean()

print(survival_by_class_and_sex)

Pclass  Sex   
1       female    0.968085
        male      0.368852
2       female    0.921053
        male      0.157407
3       female    0.500000
        male      0.135447
Name: Survived, dtype: float64


## Data Transformation

Data transformation means creating or modifying columns so that the dataset becomes more useful for analysis. In this section, we will create new features from the existing Titanic dataset: `Family_Size`, `Is_Alone`,`Fare_Per_Person`. These transformations make the dataset more suitable for analysis.


In [35]:
cleaned_df["Family_Size"] = cleaned_df["SibSp"] + cleaned_df["Parch"] + 1

cleaned_df[["SibSp", "Parch", "Family_Size"]].head()

,SibSp,Parch,Family_Size
0,1,0,2
1,1,0,2
2,0,0,1
3,1,0,2
4,0,0,1


### Identifying Passengers Travelling Alone

The `Is_Alone` column identifies whether a passenger was travelling alone. A passenger is considered alone when `Family_Size` is equal to 1.

In [36]:
cleaned_df["Is_Alone"] = cleaned_df["Family_Size"] == 1

cleaned_df[["Family_Size", "Is_Alone"]].head()

,Family_Size,Is_Alone
0,2,False
1,2,False
2,1,True
3,2,False
4,1,True


### Calculating Fare Per Person

The `Fare_Per_Person` column calculates the fare relative to the passenger's family size. This creates another useful numerical feature for analysis.

In [37]:
cleaned_df["Fare_Per_Person"] = (
    cleaned_df["Fare"] / cleaned_df["Family_Size"]
)

cleaned_df[["Fare", "Family_Size", "Fare_Per_Person"]].head()

,Fare,Family_Size,Fare_Per_Person
0,7.2500,2,3.62500
1,71.2833,2,35.64165
2,7.9250,1,7.92500
3,53.1000,2,26.55000
4,8.0500,1,8.05000


### Check the Transformed Dataset

The following code displays the new columns and confirms that the transformations were applied successfully.

In [38]:
cleaned_df[
    ["PassengerId", "Pclass", "Fare", "Family_Size",
     "Is_Alone", "Fare_Per_Person"]
].head(10)

,PassengerId,Pclass,Fare,Family_Size,Is_Alone,Fare_Per_Person
0,1,3,7.2500,2,False,3.62500
1,2,1,71.2833,2,False,35.64165
2,3,3,7.9250,1,True,7.92500
3,4,1,53.1000,2,False,26.55000
4,5,3,8.0500,1,True,8.05000
5,6,3,8.4583,1,True,8.45830
6,7,1,51.8625,1,True,51.86250
7,8,3,21.0750,5,False,4.21500
8,9,3,11.1333,3,False,3.71110
9,10,2,30.0708,2,False,15.03540


## Merging Datasets

Merging allows information from separate DataFrames to be combined using a common column. To demonstrate merging, we create a small lookup DataFrame containing the passenger class number and its corresponding class name. We then merge this information into the cleaned Titanic dataset using `Pclass`.

In [39]:
class_info = pd.DataFrame({
    "Pclass": [1, 2, 3],
    "Class_Name": ["First Class", "Second Class", "Third Class"]
})

class_info

,Pclass,Class_Name
0,1,First Class
1,2,Second Class
2,3,Third Class


### Merging 

In [40]:
cleaned_df = cleaned_df.merge(
    class_info,
    on="Pclass",
    how="left"
)

cleaned_df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Family_Size,Is_Alone,Fare_Per_Person,Class_Name
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2,False,3.62500,Third Class
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2,False,35.64165,First Class
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1,True,7.92500,Third Class
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2,False,26.55000,First Class
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1,True,8.05000,Third Class


### Verifying the merge

In [41]:
cleaned_df[["Pclass", "Class_Name"]].drop_duplicates()

,Pclass,Class_Name
0,3,Third Class
1,1,First Class
9,2,Second Class


## Final Data-Quality Check

After cleaning and transforming the dataset, we perform a final quality check.
We check:
Number of rows and columns
Missing values
Duplicate rows
Data types
Final column names

In [42]:
print("Final dataset shape:")
print(cleaned_df.shape)

print("\nMissing values:")
print(cleaned_df.isnull().sum())

print("\nDuplicate rows:")
print(cleaned_df.duplicated().sum())

print("\nData types:")
print(cleaned_df.dtypes)

Final dataset shape:
(891, 15)

Missing values:
PassengerId        0
Survived           0
Pclass             0
Name               0
Sex                0
Age                0
SibSp              0
Parch              0
Ticket             0
Fare               0
Embarked           0
Family_Size        0
Is_Alone           0
Fare_Per_Person    0
Class_Name         0
dtype: int64

Duplicate rows:
0

Data types:
PassengerId          int64
Survived             int64
Pclass               int64
Name                object
Sex                 object
Age                float64
SibSp                int64
Parch                int64
Ticket              object
Fare               float64
Embarked            object
Family_Size          int64
Is_Alone              bool
Fare_Per_Person    float64
Class_Name          object
dtype: object


In [43]:
print("Columns in final dataset:")
print(cleaned_df.columns.tolist())

Columns in final dataset:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Embarked', 'Family_Size', 'Is_Alone', 'Fare_Per_Person', 'Class_Name']


#### 
since here the cabin is missing while handling the missing values so gtry to add the cabin back

In [44]:
cleaned_df["Cabin"] = cleaned_df["Cabin"].fillna("Unknown")

KeyError: 'Cabin'

In [45]:
print("Cabin in raw dataset:", "Cabin" in df_raw.columns)
print("Cabin in cleaned dataset:", "Cabin" in cleaned_df.columns)

NameError: name 'df_raw' is not defined

In [46]:
# Reload the original Titanic dataset
raw_titanic = pd.read_csv("titanic.csv")

print("Cabin in raw dataset:", "Cabin" in raw_titanic.columns)
print("Rows in raw dataset:", len(raw_titanic))

FileNotFoundError: [Errno 2] No such file or directory: 'titanic.csv'

In [47]:
# Load the original Titanic dataset from the public source
raw_titanic = pd.read_csv(
    "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
)

print("Cabin in raw dataset:", "Cabin" in raw_titanic.columns)
print("Rows in raw dataset:", len(raw_titanic))
print("Columns in raw dataset:", len(raw_titanic.columns))

Cabin in raw dataset: True
Rows in raw dataset: 891
Columns in raw dataset: 12


In [48]:
# Keep only PassengerId and Cabin from the original dataset
cabin_info = raw_titanic[["PassengerId", "Cabin"]].copy()

# Add the original Cabin information to the cleaned dataset
cleaned_df = cleaned_df.merge(
    cabin_info,
    on="PassengerId",
    how="left"
)

# Replace missing Cabin values with "Unknown"
cleaned_df["Cabin"] = cleaned_df["Cabin"].fillna("Unknown")

print("Cabin column restored successfully.")
print("Number of rows:", len(cleaned_df))
print("Missing Cabin values:", cleaned_df["Cabin"].isnull().sum())

Cabin column restored successfully.
Number of rows: 891
Missing Cabin values: 0


In [49]:
# here is the final data set check 

print("Final dataset shape:")
print(cleaned_df.shape)

print("\nColumns:")
print(cleaned_df.columns.tolist())

print("\nMissing values:")
print(cleaned_df.isnull().sum())

print("\nDuplicate rows:")
print(cleaned_df.duplicated().sum())

Final dataset shape:
(891, 16)

Columns:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Embarked', 'Family_Size', 'Is_Alone', 'Fare_Per_Person', 'Class_Name', 'Cabin']

Missing values:
PassengerId        0
Survived           0
Pclass             0
Name               0
Sex                0
Age                0
SibSp              0
Parch              0
Ticket             0
Fare               0
Embarked           0
Family_Size        0
Is_Alone           0
Fare_Per_Person    0
Class_Name         0
Cabin              0
dtype: int64

Duplicate rows:
0


### Displaying the Final dataset 

In [50]:
cleaned_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Family_Size,Is_Alone,Fare_Per_Person,Class_Name,Cabin
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2,False,3.62500,Third Class,Unknown
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2,False,35.64165,First Class,C85
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1,True,7.92500,Third Class,Unknown
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2,False,26.55000,First Class,C123
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1,True,8.05000,Third Class,Unknown
5,6,0,3,"Moran, Mr. James",male,28.0,0,0,330877,8.4583,Q,1,True,8.45830,Third Class,Unknown
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,S,1,True,51.86250,First Class,E46
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,S,5,False,4.21500,Third Class,Unknown
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S,3,False,3.71110,Third Class,Unknown
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C,2,False,15.03540,Second Class,Unknown


In [51]:
# now checking for the final verification before saving it 
print("Final dataset shape:")
print(cleaned_df.shape)

print("\nMissing values:")
print(cleaned_df.isnull().sum())

print("\nDuplicate rows:")
print(cleaned_df.duplicated().sum())

Final dataset shape:
(891, 16)

Missing values:
PassengerId        0
Survived           0
Pclass             0
Name               0
Sex                0
Age                0
SibSp              0
Parch              0
Ticket             0
Fare               0
Embarked           0
Family_Size        0
Is_Alone           0
Fare_Per_Person    0
Class_Name         0
Cabin              0
dtype: int64

Duplicate rows:
0


## Saving the titanic_clean.CSV file 

In [53]:
cleaned_df.to_csv("titanic_clean.csv", index=False)

print("Cleaned dataset saved successfully as titanic_clean.csv")

Cleaned dataset saved successfully as titanic_clean.csv


### Verifying that the file was created 

In [54]:
print("titanic_clean.csv exists:", os.path.exists("titanic_clean.csv"))

titanic_clean.csv exists: True


# Conclusion
In Week 3,I learned how to use NumPy for numerical computing and Pandas for data wrangling.
I practiced:
- Creating and indexing NumPy arrays
- Broadcasting and vectorized operations
- Loading a public CSV dataset
- Exploring dataset structure
- Identifying and handling missing values
- Removing duplicate rows
- Filtering rows
- Grouping data using `groupby()`
- Merging DataFrames
- Creating new features
- Performing final data-quality checks
- Saving an analysis-ready dataset
The Titanic dataset was transformed from a raw dataset into a cleaned and analysis-ready dataset.
The final cleaned dataset contains 891 rows and 16 columns, with no missing values and no duplicate rows.
The cleaned dataset was saved as `titanic_clean.csv`, and a reusable `data_cleaning.py` script was created to automate the cleaning process.

# Week 3 Evidence

The Week 3 requirements were completed as follows:

-  NumPy array creation
-  NumPy indexing
-  NumPy broadcasting
-  NumPy vectorized operations
-  Public CSV dataset loaded
-  Dataset structure explored
-  Missing values identified
-  Missing values handled
-  Duplicate rows checked and removed
-  Rows filtered
-  GroupBy operations performed
-  Data transformation performed
-  DataFrames merged
-  Final data-quality check performed
-  Cleaned dataset saved
-  Data-cleaning script created
-  Data-cleaning script tested successfully
-  Notebook conclusion added